AEPServices/Sources/storage/FileSystemNamedCollection.swift

```diff
@@ -0,0 +1,141 @@
+        queue.async {
+            guard let fileUrl = self.fileUrl(for: collectionName) else {
+                return
+            }
+            if var dict = self.getDictFor(collectionName: collectionName) {
+                dict.removeValue(forKey: key)
+                if let updatedStorageData = try? JSONSerialization.data(withJSONObject: dict) {
+                    do {
+                        try updatedStorageData.write(to: fileUrl, options: .atomic)
+                    } catch {
+                        Log.error(label: self.LOG_TAG, "Error when attempting to remove value from file: \(error)")
+                    }
+                }
+            }
+        }
+    }
+
+    ///
+    /// Gets the JSON dictionary from the file with the given collection name
+    /// - Parameter collectionName: The collectionName, in this case used to destinguish a file
+    /// - Returns: The JSON dictionary if it exists, or nil if there is an error / it doesn't exist
+    ///
+    private func getDictFor(collectionName: String) -> [String: Any]? {
+        guard let fileUrl = fileUrl(for: collectionName) else {
+            return nil
+        }
+
+        if let storageData = try? Data(contentsOf: fileUrl) {
+            if let jsonResult = try? JSONSerialization.jsonObject(with: storageData) as? [String: Any] {
+                return jsonResult
+            }
+        }
+
+        return nil
+    }
+
+    ///
+    /// Gets the URL for the file given a collection name
+    /// - Parameter collectionName: The collectionName, in this case is used to destinguish a file
+    /// - Returns: The URL to the file or nil if it could not be found
+    ///
+    private func fileUrl(for collectionName: String) -> URL? {
+        if let appGroupUrl = appGroupUrl {
+            return findOrCreateAdobeSubdirectory(at: appGroupUrl)?.appendingPathComponent(collectionName).appendingPathExtension("json")
+        } else {
+            let filePath = fileManager.urls(for: .libraryDirectory, in: .allDomainsMask)[0]
+            return findOrCreateAdobeSubdirectory(at: filePath)?.appendingPathComponent(collectionName).appendingPathExtension("json")
+        }
+    }
+
+    ///
+    /// Finds or creates the Adobe subdirectory where all data store files will reside
+    /// - Parameter baseUrl: The baseUrl where the Adobe directory should reside
+    /// - Returns: The URL to the Adobe directory or nil if it could not be found or created
+    ///
+    private func findOrCreateAdobeSubdirectory(at baseUrl: URL) -> URL? {
+        // Validate baseUrl
+        if baseUrl.isSafeUrl() {
+            let adobeBaseUrl = baseUrl.appendingPathComponent(adobeDirectory, isDirectory: true)
+            do {
+                try fileManager.createDirectory(at: adobeBaseUrl, withIntermediateDirectories: true)
+            } catch {
+                Log.error(label: adobeDirectory, "Failed to create storage directory with error: \(error)")
```

Comment:
Can you check if directory exists and exit before attempting to create a directory?

Reply:
Apparently Apple doesn't recommend that approach. https://developer.apple.com/documentation/foundation/filemanager/1410277-fileexists

If you see the note towards the bottom which says: "Attempting to predicate behavior based on the current state of the file system or a particular file on the file system is not recommended. Doing so can cause odd behavior or race conditions. It’s far better to attempt an operation (such as loading a file or creating a directory), check for errors, and handle those errors gracefully than it is to try to figure out ahead of time whether the operation will succeed."

I have tested this function, and if the directory exists, it doesn't do anything. So in our case it works just fine.